File: medium/maxSubArray/maxSubArray.ts
//npx ts-node maxSubArray.ts
// Given an integer array nums, find the 
// subarray
// with the largest sum, and return its sum.

function maxSubArray(nums: number[]): number {

    let largestSum = nums[0];

    // start with the first number and add each num from array. (two for loops)
    for (let i = 0; i < nums.length; i++) {
        let currentSum = nums[i];
        if (nums[i] > largestSum) {
            largestSum = nums[i]
        }

        for (let j = i + 1; j < nums.length; j++) {
            currentSum += nums[j]

            if (currentSum > largestSum) {
                largestSum = currentSum
            }
        }

    }
    // add the largest sum of the round to the map. 
    // take the largest key of the map and return final array
    // console.log("largerst sum so far ", largestSum, "and its array", resultMap[largestSum])
    return largestSum
}

console.log(maxSubArray([-2, 1, -3, 4, -1, 2, 1, -5, 4]))
console.log(maxSubArray([5, 4, -1, 7, 8]))
console.log(maxSubArray([-2, 1]))

// alt solution 
function maxSubArraySolution(nums: number[]): number {
    let sum = nums[0]; // current sum
    let max = nums[0]; // max sum
    let ptr = 1; // pointer

    while (ptr < nums.length) {
        sum = Math.max(nums[ptr], sum + nums[ptr]);
        max = Math.max(max, sum);
        ptr += 1;
    }

    return max;
};

File: medium/maxSubArray/maxSubArray.go
package main

import (
	"fmt"
)

func main() {
	testSlice := []int{-2, 1, -3, 4, -1, 2, 1, -5, 4}
	fmt.Println(maxSubArray(testSlice))
}

/*
to find the largest sub array, we'll need to keep track of the current sum.
current number since we'll be comparing it, and the largest sum.
we can continually compare the current and largest sum to return it
*/
func maxSubArray(nums []int) int {
	// track the current, max sum, and current pointer
	currentSum := nums[0]
	maxSum := nums[0]
	pointer := 1

	// we can do a kind of while loop in go
	for pointer < len(nums) {
		// set currentSum to the max of either the currentSum or adding additional values
		currentSum = findMax(nums[pointer], currentSum+nums[pointer])
		// concurrently, we will compare the currentSum and current max to keep the larger sum
		maxSum = findMax(currentSum, maxSum)
		// move the point or else we will get stuck in an infinite while loop!
		pointer++
	}
	return maxSum

}

// since go does not provide a built in "max" func, we'll build our own!
func findMax(a, b int) int {
	if a > b {
		return a
	} else {
		return b
	}
}
